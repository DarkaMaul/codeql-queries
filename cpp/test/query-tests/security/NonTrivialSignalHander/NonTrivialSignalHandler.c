#include "../../../include/libc/string_stubs.h"
#include "../../../include/libc/signal.h"

void transitive_call() {
    printf("UNSAFE");
}

// Signal handler function
void safe_handler(int sig) {
    if (sig == SIGALRM) {
        kill(1, SIGALRM);
    }
}

void unsafe_handler(int sig) {
    transitive_call();
}

int main() {
    // Register the signal handler
    if (signal(SIGALRM, safe_handler) == SIG_ERR) {
        perror("Unable to catch SIGALRM");
        exit(EXIT_FAILURE);
    }

    if (signal(SIGALRM, unsafe_handler) == SIG_ERR) {
        perror("Unable to catch SIGALRM");
        exit(EXIT_FAILURE);
    }

    return 0;
}