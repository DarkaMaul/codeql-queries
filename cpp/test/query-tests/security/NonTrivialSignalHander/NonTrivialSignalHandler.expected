| NonTrivialSignalHandler.c:15:6:15:19 | unsafe_handler | Signal handler registered |
